src/ranked_match.cc

```diff
@@ -207,14 +223,19 @@ bool RankedMatch::operator<(const RankedMatch& other) const
         // compare codepoints
         it1 = utf8::character_start(it1, last1);
         it2 = utf8::character_start(it2, last2);
+        const auto itsave1 = it1, itsave2 = it2;
         const auto cp1 = utf8::read_codepoint(it1, end1);
         const auto cp2 = utf8::read_codepoint(it2, end2);
         if (cp1 != cp2)
         {
-            const bool punct1 = iswpunct((wchar_t)cp1);
-            const bool punct2 = iswpunct((wchar_t)cp2);
-            if (punct1 != punct2)
-                return punct1;
+            auto itlast1 = character_start_safe(itsave1 - 1, begin1);
+            auto itlast2 = character_start_safe(itsave2 - 1, begin2);
+            const auto cplast1 = utf8::read_codepoint(itlast1, end1);
+            const auto cplast2 = utf8::read_codepoint(itlast2, end2);
```

Comment:
I think this will never consider the first char as a word boundary, as character_safe_start will trigger comparison of this character against itself. I think a better alternative would be to make `cplastX = prev_codepoint(itlastX, beginX, endX).value_or(0)` with prev_codepoint returning an Optional<Codepoint> when it exists.

Reply:
* I implemented prev_codepoint and put it into the utf8 namespace, since it seems reusable

* I added cp_zero so my code doesn't add a warning or need a `(Codepoint) 0` cast. But let me know what you prefer.

* I added

```cpp
kak_assert(RankedMatch{"b(fa", "a"} < RankedMatch{"(bfa", "a"} );
```

Because one might think, following the logic of `find(1p)` this should be sorted the opposite way, but it is essentially the same as:

```cpp
kak_assert(RankedMatch{"init", ""} < RankedMatch{"__init__", ""});
```

Which is my initial case (#3015) and it is important for completing something like `sys.` in python. At least after trying for a while I didn't find a useful criteria to distinguish these cases.

* And of course I was wrong about combining the two is_word_boundary, I stashed away `debug =? yes`  The tests would fail and its also clear that the count needs directed word-boundaries.